sentrework.tex

```diff
@@ -0,0 +1,230 @@
+single and cross-type variants of \tcode{EqualityComparable} as refinements
+thereof.
+
+\pnum
+Notably, nothing in this list of properties requires comparison of sentinels
+with other sentinels -- it's not a useful feature for generic code. The
+requirement for equality comparison of sentinels in N4560 is a checkbox feature
+to enable satisfaction of cross-type \tcode{EqualityComparable}. Absent the
+\tcode{EqualityComparable} requirement, nothing in the design requires equality
+comparison for sentinels. It's nonsensical to require sentinels to have
+\tcode{Regular} types if \tcode{==} is never required to be valid or useful,
+they should be \tcode{Semiregular}.
+
+\pnum
+The list of properties also does not require that iterator types be
+\tcode{EqualityComparable} to participate in the \tcode{Sentinel} relationship.
+Just as removing the cross-type \tcode{EqualityComparable} requirement
+makes it unnecessary for sentinels to be \tcode{EqualityComparable} with
+other sentinels, it also removes the necessity for the iterator type to be
+\tcode{EqualityComparable}. In N4560 terms, \tcode{Sentinel} should be a
+relationship between a \tcode{WeakIterator} type that denotes an element of a
+range and a \tcode{Semiregular} type that represents an end-of-range calculation
+as a predicate over \tcode{WeakIterator} values embodied in the form of the
+\tcode{==} operator, i.e.:
+\begin{codeblock}
+template <class S, class I>
+concept bool Sentinel() {
+  return Semiregular<S>() && WeakIterator<I>() &&
+    BikeshedEqualityComparable<S, I>();
+}
+\end{codeblock}
+a non-weak iterator is capable of both denoting an element and the end of a
+range. In other words, a ``strong'' iterator is a type that is simultaneously a
+weak iterator and a sentinel for that weak iterator:
+\begin{codeblock}
+template <class I>
+concept bool Iterator() {
+  return Sentinel<I, I>() &&
+    EqualityComparable<I>(); // Keep the semantic property that == means
+                             // substitutable in equality-preserving expressions
+}
+\end{codeblock}
+
+\rSec1[sentrework.ram]{Ramifications}
+
+\pnum
+Correcting the inconsistency in the \tcode{Sentinel} semantics creates a class
+of ``weak'' ranges -- ranges with \tcode{WeakInput} or \tcode{WeakOutput}
+iterator types -- that the algorithms do not admit. It would appear that
+algorithms that accept input (output) ranges should possibly be relaxed to
+accept weak input (weak output) ranges instead. How would doing so constrain
+the implementations of the algorithms? Do weak ranges have less expressive power?
+
+\rSec2[sentrework.ram.singlepass]{Single-pass Algorithms}
+
+\pnum
+Algorithms that accept input or output iterators necessarily perform at most one
+pass over the range of elements those iterators reference, N4560's weak input and
+weak output iterators are consistent with those single-pass semantics. The
+distinction is obviously support for equality comparison. How do algorithms use
```

Comment:
"distinction is..support" is complete, the "obviously" is out of place. "The sole distinction is support for equality comparison." is more direct and reads better.